src/linear_algebra/decomposition.lean

```diff
@@ -0,0 +1,263 @@
+/-
+Copyright (c) 2021 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hanting Zhang
+-/
+import linear_algebra.direct_sum_module
+
+/-!
+# Decompositions of Modules
+This file talks about decompositions of modules.
+-/
+
+universes u v w
+
+open_locale direct_sum classical big_operators
+open classical linear_map dfinsupp direct_sum
+noncomputable theory
```

Comment:
I'd be interested to know exactly which bits of this are noncomputable.

My hunch right now is that this is not a good choice of definition of `decomposition`; I think the decomposition should just hold the actual equiv, and you should have a `noncomputablr def decomposition.of_independent` that constructs one from the fields you currently have.

Reply:
I don't really care what the final definition is, as long as the same total information is still accessible. Are there benefits to holding to actual equiv? 

```lean
def decomposition (ι : Type u) (R : Type v) (M : Type w)
  [semiring R] [add_comm_monoid M] [module R M] := 
subtype (λ (p : ι → submodule R M), submodule_is_internal p)


structure decomposition' (ι : Type u) (R : Type v) (M : Type w)
  [semiring R] [add_comm_monoid M] [module R M] := 
(factors : ι → submodule R M)
(to_equiv : (⨁ i, factors i) ≃ₗ[R] M)
```

These were the two other definitions I was thinking about, but in the end they honestly look exactly the same to me :P

Reply:
To be clear by "hold the actual equiv" I assume you mean something like the second one?

Reply:
Yes, although I guess you don't want to hold the equiv, rather just the inverse and the proof it is an inverse of the function we care about.

The advantage of doing that is you can compute with decompositions.

Reply:
I can reply with a mwe on Monday, but can't really do so from mobile.